### PR TITLE
Refactor, clean up ZipFile assembly

### PR DESCRIPTION
**Pull request by ianhays**
This is a net-zero change PR to move around some files in the System.IO.Compression.ZipFile assembly in preparation for some upcoming major additions.

- `src/ZipFile` is split up into:
  - src/ZipFile.Create 
  - src/ZipFile.Extract
  - src/ZipFileUtils
- `src/ZipFileExtensions` is split up into:
  - src/ZipFileExtensions.ZipArchive.Create
  - src/ZipFileExtensions.ZipArchive.Extract
  - src/ZipFileExtensions.ZipArchiveEntryExtract
- Tests from `ZipFileConvenienceMethods`, `ZipFileConvenienceMethods.netcoreapp1.1`, `ZipFileInvalidFileTests`, and `ZipFileReadOpenUpdateTests` are split up by their operating class and moved into:
  - tests/ZipFile.Create 
  - tests/ZipFile.Extract
  - tests/ZipFileExtensions.ZipArchive.Create
  - tests/ZipFileExtensions.ZipArchive.Extract
  - tests/ZipFileExtensions.ZipArchiveEntryExtract

Note that this is not a comprehensive cleanup of the ZipFile classes/members themselves, but rather a clean up and reorganization of the files. My aim is not to make any code changes yet so that I can keep this PR simple.

PTAL: @stephentoub @ViktorHofer

### CODE REVIEW

File: src/System.IO.Compression.ZipFile/src/System/IO/Compression/ZipFile.Create.cs [hunk 1]
```diff
@@ -371,208 +351,12 @@ public static void CreateFromDirectory(string sourceDirectoryName, string destin
...
-                archive.ExtractToDirectory(destinationDirectoryName, overwrite);
-            }
-        }
+                                               CompressionLevel compressionLevel, bool includeBaseDirectory, Encoding entryNameEncoding) =>
+            DoCreateFromDirectory(sourceDirectoryName, destinationArchiveFileName, compressionLevel, includeBaseDirectory, entryNameEncoding: null);
 
         private static void DoCreateFromDirectory(string sourceDirectoryName, string destinationArchiveFileName,
-                                                  CompressionLevel? compressionLevel, bool includeBaseDirectory,
-                                                  Encoding entryNameEncoding)
+                                               CompressionLevel? compressionLevel, bool includeBaseDirectory, Encoding entryNameEncoding)
```
Comment (stephentoub) [inline src/System.IO.Compression.ZipFile/src/System/IO/Compression/ZipFile.Create.cs]: Nit: indenting?

File: src/System.IO.Compression.ZipFile/src/System/IO/Compression/ZipFile.Extract.cs [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+        /// <exception cref="NotSupportedException">sourceArchive or destinationDirectoryName is in an invalid format. </exception>
+        /// <exception cref="FileNotFoundException">sourceArchive was not found.</exception>
+        /// <exception cref="InvalidDataException">The archive specified by sourceArchive: Is not a valid ZipArchive
+        /// -or- An archive entry was not found or was corrupt. -or- An archive entry has been compressed using a compression method
+        /// that is not supported.</exception>
+        /// 
+        /// <param name="sourceArchiveFileName">The path to the archive on the file system that is to be extracted.</param>
+        /// <param name="destinationDirectoryName">The path to the directory on the file system. The directory specified must not exist, but the directory that it is contained in must exist.</param>
+        public static void ExtractToDirectory(string sourceArchiveFileName, string destinationDirectoryName) => ExtractToDirectory(sourceArchiveFileName, destinationDirectoryName, entryNameEncoding: null, overwrite: false);
```
Comment (stephentoub) [inline src/System.IO.Compression.ZipFile/src/System/IO/Compression/ZipFile.Extract.cs]: Nit: These are long lines... maybe you could split them, at least moving this to the next line and indenting.

File: src/System.IO.Compression.ZipFile/src/System/IO/Compression/ZipFileExtensions.ZipArchive.Create.cs [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        /// <param name="compressionLevel">The level of the compression (speed/memory vs. compressed size trade-off).</param>
+        /// <returns>A wrapper for the newly created entry.</returns>
+        public static ZipArchiveEntry CreateEntryFromFile(this ZipArchive destination,
+                                                          string sourceFileName, string entryName, CompressionLevel compressionLevel) =>
+            DoCreateEntryFromFile(destination, sourceFileName, entryName, compressionLevel);
+
+        internal static ZipArchiveEntry DoCreateEntryFromFile(this ZipArchive destination,
+                                                            string sourceFileName, string entryName, CompressionLevel? compressionLevel)
+        {
+                if (destination == null)
```
Comment (stephentoub) [inline src/System.IO.Compression.ZipFile/src/System/IO/Compression/ZipFileExtensions.ZipArchive.Create.cs]: Nit: indenting, here and the line above

File: src/System.IO.Compression.ZipFile/tests/ZipFile.Extract.cs [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+        public void Windows_ZipWithOSSpecificFileNames(string zipName, string fileName)
+        {
+            string tempDir = GetTestFilePath();
+            ZipFile.ExtractToDirectory(compat(zipName) + ".zip", tempDir);
+            string[] results = Directory.GetFiles(tempDir, "*", SearchOption.AllDirectories);
+            Assert.Equal(1, results.Length);
+            Assert.Equal(fileName, Path.GetFileName(results[0]));
+        }
+
+#if netcoreapp
```
Comment (ViktorHofer) [inline src/System.IO.Compression.ZipFile/tests/ZipFile.Extract.cs]: nit: put that into its separate file: `ZipFile.Extract.netcoreapp.cs`?

File: src/System.IO.Compression.ZipFile/src/System/IO/Compression/ZipFileExtensions.ZipArchive.Create.cs [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+        /// <exception cref="ObjectDisposedException">The ZipArchive has already been closed.</exception>
+        /// 
+        /// <param name="sourceFileName">The path to the file on the file system to be copied from. The path is permitted to specify relative
+        /// or absolute path information. Relative path information is interpreted as relative to the current working directory.</param>
+        /// <param name="entryName">The name of the entry to be created.</param>
+        /// <param name="compressionLevel">The level of the compression (speed/memory vs. compressed size trade-off).</param>
+        /// <returns>A wrapper for the newly created entry.</returns>
+        public static ZipArchiveEntry CreateEntryFromFile(this ZipArchive destination,
+                                                          string sourceFileName, string entryName, CompressionLevel compressionLevel) =>
+            DoCreateEntryFromFile(destination, sourceFileName, entryName, compressionLevel);
```
Comment (ViktorHofer) [inline src/System.IO.Compression.ZipFile/src/System/IO/Compression/ZipFileExtensions.ZipArchive.Create.cs]: and also this line

**Comment by stephentoub** [comment]
@dotnet-bot test OSX x64 Debug Build please ("Error during cleanup: java.lang.IllegalArgumentException: Failed to prepare archiveArtifacts step")

**Pull request merged**